### PR TITLE
[Merged by Bors] - refactor(LinearAlgebra/QuadraticForm/Basic): Use bilinear maps for the companion

### PR DESCRIPTION
**Pull request by mans0954**
This PR replaces the companion `(B : BilinForm R M)` of a quadratic form with a bilinear map `(B : M →ₗ[R] M →ₗ[R] R)`.

This is intended as a baby step towards generalising quadratic forms to quadratic maps. In the future we could allow for companion bilinear or even sesquilinear maps into an `R`-module.

---

I've previously tried to go straight to quadratic maps (#7569) but it's making for a PR that is too big and unwieldy. This PR is hopefully more achievable.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mans0954** [comment]
!bench

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean [hunk 1]
```diff
@@ -352,20 +358,19 @@ def ofPolar (toFun : M → R) (toFun_smul : ∀ (a : R) (x : M), toFun (a • x)
...
-          bilin_add_right := fun x y z => by simp_rw [polar_comm _ x, polar_add_left]
-          bilin_smul_right := fun r x y => by
-            simp_rw [polar_comm _ x, polar_smul_left, smul_eq_mul] },
-        fun x y => by rw [BilinForm.coeFn_mk, polar, sub_sub, add_sub_cancel'_right]⟩ }
+    exists_companion' := ⟨LinearMap.mk₂ R (polar toFun) (polar_add_left) (polar_smul_left)
+      (fun x _ _ => by simp_rw [polar_comm _ x, polar_add_left])
+      (fun _ _ _ => by rw [polar_comm, polar_smul_left, polar_comm]),
+      fun _ _ =>  by
+      simp only [LinearMap.mk₂_apply]
+      rw [polar, sub_sub, add_sub_cancel'_right]⟩}
```
Comment (mcdoll) [inline Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:366]:
```suggestion
        simp only [LinearMap.mk₂_apply]
        rw [polar, sub_sub, add_sub_cancel'_right]⟩ }
```

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean [hunk 2]
```diff
@@ -654,13 +663,19 @@ section Semiring
...
+
 variable {B : BilinForm R M}
 
 /-- A bilinear form gives a quadratic form by applying the argument twice. -/
-def toQuadraticForm (B : BilinForm R M) : QuadraticForm R M where
-  toFun x := B x x
-  toFun_smul a x := by simp only [mul_assoc, smul_right, smul_left]
-  exists_companion' := ⟨B + BilinForm.flipHom ℕ B, fun x y => by simp [add_add_add_comm, add_comm]⟩
+def toQuadraticForm (B : BilinForm R M) : QuadraticForm R M :=
+  LinearMap.toQuadraticForm (BilinForm.toLin B)
```
Comment (mcdoll) [inline Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:678]: untested:
```suggestion
def toQuadraticForm (B : BilinForm R M) : QuadraticForm R M :=
  B.toLin.toQuadraticForm
```

File: Mathlib/LinearAlgebra/QuadraticForm/Dual.lean [hunk 3]
```diff
@@ -52,14 +51,15 @@ section Ring
 variable [CommRing R] [AddCommGroup M] [Module R M]
 
 theorem nondenerate_dualProd :
-    (dualProd R M).Nondegenerate ↔ Function.Injective (Module.Dual.eval R M) := by
+    (LinearMap.toBilin (dualProd R M)).Nondegenerate
+    ↔ Function.Injective (Module.Dual.eval R M) := by
```
Comment (mcdoll) [inline Mathlib/LinearAlgebra/QuadraticForm/Dual.lean:55]: There is [LinearMap.SeparatingLeft](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/SesquilinearForm.html#LinearMap.SeparatingLeft) which is the same as `BilinForm.Nondegenerate`.

**Review by mcdoll — COMMENTED** [review]
Thank you so much, I would be very happy if we could remove `BilinForm` entirely from mathlib and this is a big step.

File: Mathlib/LinearAlgebra/QuadraticForm/Dual.lean [hunk 4]
```diff
@@ -34,8 +34,7 @@ variable [CommSemiring R] [AddCommMonoid M] [Module R M]
 /-- The symmetric bilinear form on `Module.Dual R M × M` defined as
 `B (f, x) (g, y) = f y + g x`. -/
 @[simps!]
-def dualProd : BilinForm R (Module.Dual R M × M) :=
-  LinearMap.toBilin <|
+def dualProd :=
```
Comment (mcdoll) [inline Mathlib/LinearAlgebra/QuadraticForm/Dual.lean:37]: Please add the type annotation

Comment (mans0954) [reply Mathlib/LinearAlgebra/QuadraticForm/Dual.lean]: Done, thanks.

File: Mathlib/LinearAlgebra/QuadraticForm/Dual.lean (hunk 3 above)
Comment (mans0954) [reply Mathlib/LinearAlgebra/QuadraticForm/Dual.lean]: That's better, thanks.

File: Mathlib/LinearAlgebra/QuadraticForm/Dual.lean [hunk 5]
```diff
@@ -52,14 +51,13 @@ section Ring
 variable [CommRing R] [AddCommGroup M] [Module R M]
 
 theorem nondenerate_dualProd :
```
Comment (mcdoll) [inline Mathlib/LinearAlgebra/QuadraticForm/Dual.lean:53]:
```suggestion
theorem separatingLeft_dualProd :
```
and probably there is an #align that has to be changed as well

**Comment by mans0954** [comment]
!bench

**Comment by mcdoll** [comment]
maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean [hunk 6]
```diff
@@ -377,7 +382,7 @@ variable [CommSemiring R] [AddCommMonoid M] [Module R M]
 section SMul
 
 variable [Monoid S] [Monoid T] [DistribMulAction S R] [DistribMulAction T R]
-variable [SMulCommClass S R R] [SMulCommClass T R R]
+variable [SMulCommClass R S R] [SMulCommClass S R R] [SMulCommClass T R R] [SMulCommClass R T R]
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:385]: This looks like a regression to me; one implies the other so there should be no need for both

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:385]: Addressed in #10167

**Pull request closed without merging**